### PR TITLE
[InstCombine] Fold (sub (xor X, (sext C)), (sext C)) => (select C (neg X), X)

### PR DESCRIPTION
**Pull request by bzEq**
This is useful when computing absdiff.

Correctness prove: https://alive2.llvm.org/ce/z/eMbxps, https://alive2.llvm.org/ce/z/SNCWJe.

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2456]:
```suggestion
        match(Op1, m_SExt(m_Specific(C0)))) {
```

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2454]: Please also handle the commuted version: `(sub (sext C), (xor X, (sext C))) -> select C, X, (neg X)`.

Missing one-use checks?

File: llvm/test/CodeGen/AArch64/absdiff.ll
Comment (dtcxzyw) [inline llvm/test/CodeGen/AArch64/absdiff.ll:1]: Don't add tests in `CodeGen`.

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:4]: What is the motivation of this patch? Does this pattern exist in some real-world applications?

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (bzEq) [reply llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:4]: > Does this pattern exist in some real-world applications?

Yes. We have observed this pattern in our internal workload.

Comment (bzEq) [reply llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:4]: Similar to https://www.oreilly.com/library/view/hackers-delight/0201914654/0201914654_ch02lev1sec4.html.

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (bzEq) [inline llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:12]: It looks like `m_Xor` isn't commutative aware, so instcombine fails to catch the pattern.
`m_Xor` is implemented as
```
template <typename LHS, typename RHS>
inline BinaryOp_match<LHS, RHS, Instruction::Xor> m_Xor(const LHS &L,
                                                        const RHS &R) {
  return BinaryOp_match<LHS, RHS, Instruction::Xor>(L, R);
}
```
Does anybody know why isn't `Commutable` set `true`?

Comment (bzEq) [reply llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:12]: Thanks @chenzheng1030 pointed out offline, we can use `m_c_Xor` here.

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (dtcxzyw) [inline llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:29]: Please add some negative tests:
+ mismatched type: (sub (xor X, (sext i8 C)), (sext i8 C))
+ mismatched C: (sub (xor X, (sext C1)), (sext C2))
+ (xor X, (sext C)) has multiple uses
+ (sext C) has multiple uses

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2456]: What is the purpose of the m_OneUse?

Could the patterns be better matched as?
```cpp
return match(LHS, m_c_Xor(m_Value(X), m_Specific(RHS))) &&
       match(RHS, m_SExt(m_Specific(C)) &&
       (C->getType()->getScalarSizeInBits() == 1);
```

Comment (bzEq) [reply llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp]: > What is the purpose of the m_OneUse?

Looks no need to check `m_OneUse` of the `i1` value, since we are still using it in the result.
What concerns me is usage of `(sext i1)`, if it's still using by other instructions other than this `(sub (sext C), (xor X, (sext C)))`, not sure if it's still profitable. I'll add test cases to demonstrate as suggested by @dtcxzyw .

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (bzEq) [inline llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:98]: Looks we should limit one use of `(xor X, (sext C))` to avoid regression.

**Review by dtcxzyw — APPROVED** [review]
LGTM. Please wait for additional approval from other reviewers.

File: llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp
Comment (RKSimon) [inline llvm/lib/Transforms/InstCombine/InstCombineAddSub.cpp:2457]: Is this better? `m_OneUse(m_c_Xor(m_Value(X), m_Specific(RHS)))`

**Review by RKSimon — APPROVED** [review]
LGTM - cheers

File: llvm/test/Transforms/InstCombine/sub-xor-cmp.ll
Comment (nikic) [inline llvm/test/Transforms/InstCombine/sub-xor-cmp.ll:28]: Unless I overlooked it, you're missing a commuted test for the xor operands.

**Pull request merged**
